inductor: fix size and stride comparison

chunyuan-w opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #94481

We met a case where `old.get_stride()` is a `tuple`: `(1, 16)` while `new.get_stride()` is a `list`: `[1, 16]`.
`old.get_stride() == new.get_stride()` returns `False` though they're actually equal.


cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

chunyuan-w committed:
inductor: fix stride comparison

[ghstack-poisoned]

EikanWang commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -36,13 +36,16 @@ def make_buffer_reuse(old, new, del_func, declare, ending, as_strided):
         del_line = del_func(old.get_name())
-    if old.get_size() == new.get_size() and old.get_stride() == new.get_stride():
+
+    old_stride = list(old.get_stride())
+    new_stride = list(new.get_stride())
```
Will `get_size` also encounter this situation? I'm wondering if we also should apply this conversion to `get_size`

chunyuan-w replied:
Made the same change for `get_size` as well.

jgong5 replied:
The `get_size()` and `get_stride()` return the size and stride of the layout and the size hint shows that it should always be a list. May I know what code initializes them to a tuple?

chunyuan-w replied:
https://github.com/pytorch/pytorch/blob/89df0e425310c8f5210921b01a24cefc2e62e541/torch/_inductor/ir.py#L3692

`make_contiguous_strides_for` here returns a `tuple`:
https://github.com/pytorch/pytorch/blob/89df0e425310c8f5210921b01a24cefc2e62e541/torch/_prims_common/__init__.py#L1374-L1376

Below functions also return `tuple` for `stride`:
  `make_channels_last_2d_strides_for`
  `make_channels_last_3d_strides_for`
  `make_channels_last_strides_for`

chunyuan-w replied:
@desertfire  I've changed the `get_size` and `get_stride` functions to return `list`. Does that look good to you?

chunyuan-w committed:
Update on "inductor: fix stride comparison"


`get_size()` returns `torch.Size()` while `get_stride()` returns `list` or `tuple`.

We met a case where `old.get_stride()` is a `tuple`: `(1, 16)` while `new.get_stride()` is a `list`: `[1, 16]`.
`old.get_stride() == new.get_stride()` returns `False` though they're actually equal.


cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

chunyuan-w committed:
Update on "inductor: fix size and stride comparison"


We met a case where `old.get_stride()` is a `tuple`: `(1, 16)` while `new.get_stride()` is a `list`: `[1, 16]`.
`old.get_stride() == new.get_stride()` returns `False` though they're actually equal.


cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

chunyuan-w committed:
Update on "inductor: fix size and stride comparison"


We met a case where `old.get_stride()` is a `tuple`: `(1, 16)` while `new.get_stride()` is a `list`: `[1, 16]`.
`old.get_stride() == new.get_stride()` returns `False` though they're actually equal.


cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

chunyuan-w committed:
Update on "inductor: fix size and stride comparison"


We met a case where `old.get_stride()` is a `tuple`: `(1, 16)` while `new.get_stride()` is a `list`: `[1, 16]`.
`old.get_stride() == new.get_stride()` returns `False` though they're actually equal.


cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

chunyuan-w committed:
Update on "inductor: fix size and stride comparison"


We met a case where `old.get_stride()` is a `tuple`: `(1, 16)` while `new.get_stride()` is a `list`: `[1, 16]`.
`old.get_stride() == new.get_stride()` returns `False` though they're actually equal.


cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

chunyuan-w committed:
Update on "inductor: fix size and stride comparison"


We met a case where `old.get_stride()` is a `tuple`: `(1, 16)` while `new.get_stride()` is a `list`: `[1, 16]`.
`old.get_stride() == new.get_stride()` returns `False` though they're actually equal.


cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

chunyuan-w committed:
Update on "inductor: fix size and stride comparison"


We met a case where `old.get_stride()` is a `tuple`: `(1, 16)` while `new.get_stride()` is a `list`: `[1, 16]`.
`old.get_stride() == new.get_stride()` returns `False` though they're actually equal.


cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

chunyuan-w committed:
Update on "inductor: fix size and stride comparison"


We met a case where `old.get_stride()` is a `tuple`: `(1, 16)` while `new.get_stride()` is a `list`: `[1, 16]`.
`old.get_stride() == new.get_stride()` returns `False` though they're actually equal.


cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

chunyuan-w commented:
@pytorchbot merge
